arithmetization/src/main/java/net/consensys/linea/sequencer/txselection/selectors/MaxBlockSizeTransactionSelector.java

```diff
@@ -0,0 +1,114 @@
+import org.hyperledger.besu.datatypes.Transaction;
+import org.hyperledger.besu.evm.log.Log;
+import org.hyperledger.besu.plugin.data.TransactionProcessingResult;
+import org.hyperledger.besu.plugin.data.TransactionSelectionResult;
+import org.hyperledger.besu.plugin.services.txselection.PluginTransactionSelector;
+
+@Getter
+@Slf4j
+public class MaxBlockSizeTransactionSelector implements PluginTransactionSelector {
+  private int cumulativeBlockBytesSize = EMPTY_L1_BLOCK_SIZE;
+  private final int maxBytesPerBlock;
+  public static final int AVERAGE_TX_RLP_SIZE = 400;
+  public static final int FROM_ADDRESS_SIZE = 20;
+  public static final int EMPTY_L1_BLOCK_SIZE = 64 + 63 + 64 + 64;
+  public static final int LOG_SIZE = 32;
+
+  public MaxBlockSizeTransactionSelector(int maxBytesPerBlock) {
+    this.maxBytesPerBlock = maxBytesPerBlock;
+  }
+
+  @Override
+  public TransactionSelectionResult evaluateTransactionPreProcessing(
+      PendingTransaction pendingTransaction) {
+    return TransactionSelectionResult.SELECTED;
+  }
+
+  @Override
+  public TransactionSelectionResult evaluateTransactionPostProcessing(
+      final PendingTransaction pendingTransaction,
+      final TransactionProcessingResult processingResult) {
+
+    int transactionL1Size =
+        calculateL1TransactionSize(pendingTransaction, processingResult.getLogs());
+    if (isTransactionExceedingL1BlockSizeLimit(transactionL1Size)) {
+      log.atTrace()
+          .setMessage(
+              "tx {} size {} would cause block (cumulative size {}) to exceed max {} bytes, skipping tx")
+          .addArgument(pendingTransaction.getTransaction()::getHash)
+          .addArgument(() -> transactionL1Size)
+          .addArgument(() -> cumulativeBlockBytesSize)
+          .addArgument(maxBytesPerBlock)
+          .log();
+      return BLOCK_BYTES_SIZE_OVERFLOW;
+    }
+    return TransactionSelectionResult.SELECTED;
+  }
+
+  /**
+   * Checks if the total size of all transactions in a block would exceed the maximum allowed size
+   * if the given transaction were added.
+   *
+   * @param L1TransactionSize the L1 Transaction Size in Bytes.
+   * @return true if the total call data size would be too big, false otherwise.
+   */
+  private boolean isTransactionExceedingL1BlockSizeLimit(int L1TransactionSize) {
+    try {
+      return Math.addExact(cumulativeBlockBytesSize, L1TransactionSize) > maxBytesPerBlock;
+    } catch (final ArithmeticException e) {
+      // Overflow won't occur as totalBlockSize won't exceed Integer.MAX_VALUE
+      return true;
+    }
+  }
+
+  /**
+   * Updates the total size when a transaction is selected.
+   *
+   * @param pendingTransaction The selected transaction.
+   */
+  @Override
+  public void onTransactionSelected(
+      final PendingTransaction pendingTransaction, final TransactionProcessingResult result) {
+    final int transactionL1Size = calculateL1TransactionSize(pendingTransaction, result.getLogs());
```

Comment:
we should not have to calculate the size here again, we have done that in the post processing already

Reply:
do you suggest saving that size so we can reuse it?

Reply:
@pinges